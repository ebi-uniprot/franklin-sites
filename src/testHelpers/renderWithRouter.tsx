import { ReactElement, ReactNode } from 'react';
import { MemoryRouter } from 'react-router-dom';
// eslint-disable-next-line import/no-extraneous-dependencies
import { render } from '@testing-library/react';

export default (ui: ReactElement) => {
  const Wrapper = ({ children }: { children: ReactNode }) => (
    <MemoryRouter>{children}</MemoryRouter>
  );
  return {
    ...render(ui, { wrapper: Wrapper }),
  };
};
