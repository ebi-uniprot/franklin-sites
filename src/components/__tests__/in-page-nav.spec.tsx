/* eslint-disable class-methods-use-this, lines-between-class-members */
import { screen, waitFor } from '@testing-library/react';

import renderWithRouter from '../../testHelpers/renderWithRouter';

import InPageNav from '../in-page-nav';

describe('InPageNav component', () => {
  beforeAll(() => {
    global.IntersectionObserver = class FakeIntersectionObserver {
      root = null;
      rootMargin = '0px';
      thresholds = [];
      observe() {
        return null;
      }
      unobserve() {
        return null;
      }
      disconnect() {
        return null;
      }
      takeRecords() {
        return [];
      }
    };
  });

  const sections = [
    {
      id: 'id1',
      label: 'First link',
    },
    {
      id: 'id2',
      label: 'Second link',
    },
    {
      id: 'id3',
      label: 'Third link',
      disabled: true,
    },
  ];

  it('should render', () => {
    const { asFragment } = renderWithRouter(<InPageNav sections={sections} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should scroll the default target element into view', async () => {
    renderWithRouter(
      <div>
        <InPageNav sections={sections} />
        <div id="id3" data-testid="target" />
      </div>,
      { route: '/#id3' }
    );
    const target = screen.getByTestId('target');
    target.scrollIntoView = jest.fn();
    await waitFor(() => expect(target.scrollIntoView).toHaveBeenCalledTimes(1));
  });

  it('should scroll the target element into view after hash change', async () => {
    const { history } = renderWithRouter(
      <div>
        <InPageNav sections={sections} />
        <div id="id3" data-testid="target" />
      </div>
    );

    const target = screen.getByTestId('target');
    target.scrollIntoView = jest.fn();

    history.push({ ...history.location, hash: '#id3' });

    await waitFor(() => expect(target.scrollIntoView).toHaveBeenCalledTimes(1));
  });

  it('should scroll to top of rootElement after hash removed', async () => {
    const { history } = renderWithRouter(
      <div>
        <InPageNav sections={sections} rootElement={document.body} />
        <div id="id3" data-testid="target" />
      </div>,
      { route: '/#id3' }
    );

    document.body.scrollTo = jest.fn();

    history.push({ ...history.location, hash: '' });

    await waitFor(() =>
      expect(document.body.scrollTo).toHaveBeenCalledTimes(1)
    );
  });
});
