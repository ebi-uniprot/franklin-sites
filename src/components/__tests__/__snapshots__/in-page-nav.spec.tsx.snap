// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InPageNav component should render 1`] = `
<DocumentFragment>
  <ul
    class="in-page-nav"
  >
    <div
      class="marker"
    />
    <li
      class=""
    >
      <a
        class="active"
        href="/#id1"
      >
        First link
      </a>
    </li>
    <li
      class=""
    >
      <a
        class=""
        href="/#id2"
      >
        Second link
      </a>
    </li>
    <li
      class="disabled"
    >
      <a
        class=""
        href="/#id3"
      >
        Third link
      </a>
    </li>
  </ul>
</DocumentFragment>
`;
